Ex 01 rest api - User Management Module

safwanone2n opened the pull request:
- Added CRUD endpoints for user managment module
-  Implemented testcases for all user CRUD endpoints
- Added logs throughout the endpoints
- Readme file with setting up and neccessery details about the application
- Utilized sqlc for type-safe queries

safwanone2n committed:
EX-01 - created directory structure, initial configs and handler structure for rest api

safwanone2n committed:
EX-01 - implemented CRUD endpoints for user module with layered architecture to segragate repo, buisiness and handler logics

safwanone2n committed:
EX-01 - add endpoint tests and structured logging

- Added table-driven tests for all user endpoints
- Introduced structured, level-based logging across handlers
- Restructured handler registration for better modularity
- Added configuration management via environment variables

safwanone2n committed:
EX-01 - updated readme file and gitignore

safwanone2n committed:
EX-01 - updated readme to include details of sqlc

safwanone2n committed:
EX-01 - added postman collection of the apis, updated readme and minor changes in user handler apis

safwanone2n committed:
EX-02 - containarizing rest api

- containarized rest api utilzing multistage build
- updated make file for run with docker image
- updated readme
- updated env collection logic to support runtime injection through
  docker

Nilesh2000 commented on REST-API/cmd/server/main.go:
```diff
@@ -0,0 +1,79 @@
+	userUseCaseI := usecases.NewUserUseCases(userRepoI)
+	userHandlerI := handlers.NewUserHandler(userUseCaseI, logger)
+
+	//initialize repo
```
what repo are you initializing for?

safwanone2n replied:
userRepo is initialized for utilization in usecases, couldn't understand context of the comment.

Nilesh2000 commented on REST-API/cmd/server/main.go:
```diff
@@ -0,0 +1,79 @@
+
+	//initializing user repo
+	userRepoI := userRepo.NewUserRepo(querier)
+	userUseCaseI := usecases.NewUserUseCases(userRepoI)
```
Isn't the convention to use `service` instead of `usecase`? Or is this a convention I am unaware of?

safwanone2n replied:
usually this term is aligning with clean arch, which i have implemented. But i understand as go convention service is correct term.  Changes will be done.

Nilesh2000 commented on REST-API/config/config.go:
```diff
@@ -0,0 +1,10 @@
+
+import "github.com/joho/godotenv"
+
+func init() {
```
bad practice to use `init` methods in golang code unless you want to execute something implicitly (which is almost never). Why not use dependency injection and then initialize everything in `main.go`?

safwanone2n replied:
This code is unused at  later commits and this will be changed to return config for DI

Nilesh2000 commented on REST-API/db/migrations/001_create_users.sql:
```diff
@@ -0,0 +1,13 @@
+-- This is a sample migration.
+
+create table users (
```
Not a comment per se, but consider having all reserve keywords in uppercase for SQL Statements. Improves readability and is a widely followed convention.

safwanone2n replied:
changes will be done.

Nilesh2000 commented on REST-API/db/queries/users.sql:
```diff
@@ -0,0 +1,38 @@
+
```
Format the file. Why are there unnecessary empty lines?
Your code should also be glaceable, besides just being functional.

safwanone2n replied:
changes will be done.

Nilesh2000 commented on REST-API/db/db.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+func GetConn(ctx context.Context) (*pgxpool.Pool, error) {
+	pgx, err := pgxpool.New(ctx, os.Getenv("DATABASE_URL"))
```
If you had used a dependency injection pattern (which is a good practice), you would only have to load the `DATABASE_URL` from the config, rather than fetching it from the environment variable like you have done above.

safwanone2n replied:
Code is in dependency injection pattern except for configs. changes will be done for this part also

Nilesh2000 commented on REST-API/internal/user/handlers/user_handler.go:
```diff
@@ -0,0 +1,308 @@
+	commonutils.SendResponse(ctx, http.StatusCreated, models.IdResponse{Id: user.Id}, nil)
+}
+
+// DeleteUserHandler implements UserHandlerI.
```
This `godoc` comment is slightly misleading. The `UserHandler` type implements the `UserHandlerI` interface. The method does not implement the interface.

safwanone2n replied:
This is go provided auto command when a constructor is created to return struct with interface. but changes will be done

Nilesh2000 commented on REST-API/internal/user/usecases/users_usecaseses.go:
```diff
@@ -0,0 +1,94 @@
+	return nil
+}
+
+func NewUserUseCases(repo userRepo.UserRepoI) UserUseCasesI {
```
Constructor is usually defined immediately after the definition of the type. So, I would suggest moving it upwards.

safwanone2n replied:
changes will be done.

Nilesh2000 reviewed (commented):
### What is good?
- Code seems to be functional.
- You have implemented tests and mocks.
- You have implemented the `Accept Interfaces, Return Structs` principle correctly.

### What can be improved?
- Code is not glanceable and not easy to make sense of. Set up a formatter and linter on pre-commit
- `godoc` comments are slightly misleading and not present at all places.
- Use Dependency Injection Pattern for loading environment variables.
- Although the code has tests, I don't know if they pass. Attempt to write a simple GitHub action or integrate a third-party tool that displays if your tests are passing on the `README`
- Test Coverage seems pretty low at the moment. Try to get 60% coverage.
- The use of the term `UseCase` is unconventional and is something I haven't come across. I suggest sticking to the handler / controller, service, repository convention.
- Follow file naming conventions. The first letter in file names `dockerfile`, `makefile` should be uppercase. Similarly, `readme.md` should be saved as `README.md`.
- Implement Graceful shutdown by checking for termination signals.

### Suggested Reading
- https://lets-go.alexedwards.net/
- https://github.com/golang-standards/project-layout

Nilesh2000 commented on REST-API/cmd/server/main.go:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	//initializing user repo
+	userRepoI := userRepo.NewUserRepo(querier)
```
What is the `I` suffix for?

safwanone2n replied:
I is utilized for readability because for knowing the returned type is Interface. Sort of my way of doing things, if not aligining with standards provide more context and suggestions.

Nilesh2000 replied:
The return type of the constructor is not an interface. It is a concrete type. So, the `I` suffix here is misleading.

safwanone2n committed:
EX_01 - refactored code to include graceful shutdown and new changes
- added dependency injection for env vars
- usecases term is refactored to service
- refactored testcases for coverage above 60 percent
- refactoring accross codebase with linters

safwanone2n committed:
EX-01 - added testcase workflow file

safwanone2n committed:
Moving GitHub Actions workflow to repo root

safwanone2n committed:
EX-01 - fixed path issue while running tests
